File: Android/app/src/main/java/ch/pfvr/internapp/RhineNavigation.java
package ch.pfvr.internapp;

/** Official Rhine navigation stages derived from Basel-Rheinhalle gauge levels. */
final class RhineNavigation {
    enum Stage { UNKNOWN, NORMAL, HWM_I, HWM_IIB, HWM_IIA }

    static final double HWM_I_CM = 700.0d;
    static final double HWM_IIB_CM = 790.0d;
    static final double HWM_IIA_CM = 820.0d;
    static final long MAX_CURRENT_AGE_MS = 60L * 60L * 1000L;

    private RhineNavigation() {}

    static Stage fromBaselGaugeCm(double gaugeCm){
        if(!Double.isFinite(gaugeCm))return Stage.UNKNOWN;
        if(gaugeCm>=HWM_IIA_CM)return Stage.HWM_IIA;
        if(gaugeCm>=HWM_IIB_CM)return Stage.HWM_IIB;
        if(gaugeCm>=HWM_I_CM)return Stage.HWM_I;
        return Stage.NORMAL;
    }

    static boolean isCurrent(long measurementTimestamp,long cacheUpdatedTimestamp,long now){
        if(measurementTimestamp<=0L||cacheUpdatedTimestamp<=0L||now<=0L)return false;
        long measurementAge=now-measurementTimestamp;
        long cacheAge=now-cacheUpdatedTimestamp;
        if(measurementAge<0L||cacheAge<0L)return false;
        return measurementAge<=MAX_CURRENT_AGE_MS&&cacheAge<=MAX_CURRENT_AGE_MS;
    }

    static Stage fromCurrentBaselGaugeCm(double gaugeCm,long measurementTimestamp,long cacheUpdatedTimestamp,long now){
        if(!isCurrent(measurementTimestamp,cacheUpdatedTimestamp,now))return Stage.UNKNOWN;
        return fromBaselGaugeCm(gaugeCm);
    }

    static String shortLabel(Stage stage){
        if(stage==Stage.UNKNOWN)return "Lage unklar";
        if(stage==Stage.HWM_I)return "HWM I";
        if(stage==Stage.HWM_IIB)return "Sperre IIb";
        if(stage==Stage.HWM_IIA)return "Sperre IIa";
        return "Normal";
    }

    static String detail(Stage stage){
        if(stage==Stage.UNKNOWN)return "Basel-Pegel fehlt oder ist älter als 60 Minuten. Massgebend sind die Schweizerischen Rheinhäfen.";
        if(stage==Stage.HWM_I)return "Voralarm ab 700 cm Pegel Basel-Rheinhalle.";
        if(stage==Stage.HWM_IIB)return "Kleinschifffahrt und Fähren Basel–Rheinfelden gesperrt.";
        if(stage==Stage.HWM_IIA)return "Schifffahrt Rheinfelden–Kembs gesperrt.";
        return "Unter Hochwassermarke I (< 700 cm).";
    }

    static Stage[] officialThresholdStages(){
        return new Stage[]{Stage.HWM_I,Stage.HWM_IIB,Stage.HWM_IIA};
    }

    static double thresholdGaugeCm(Stage stage){
        if(stage==Stage.HWM_I)return HWM_I_CM;
        if(stage==Stage.HWM_IIB)return HWM_IIB_CM;
        if(stage==Stage.HWM_IIA)return HWM_IIA_CM;
        return Double.NaN;
    }

    static double thresholdGraphValue(Stage stage,boolean centimetres){
        double cm=thresholdGaugeCm(stage);
        if(!Double.isFinite(cm))return Double.NaN;
        return centimetres?cm:240.0d+cm/100.0d;
    }

    static String thresholdMarker(Stage stage,boolean centimetres){
        double value=thresholdGraphValue(stage,centimetres);
        String prefix=stage==Stage.HWM_I?"I":(stage==Stage.HWM_IIB?"IIb":"IIa");
        return centimetres
                ?String.format(java.util.Locale.GERMAN,"%s %.0f",prefix,value)
                :String.format(java.util.Locale.GERMAN,"%s %.2f",prefix,value);
    }
}

File: iOS/Sources/PFVRCore/RhineNavigation.swift
import Foundation

/// Ports Android's documented thresholds; this is an orientation derived from BAFU raw data.
public enum RhineNavigation {
    public enum Stage: String, CaseIterable, Codable, Sendable {
        case unknown, normal, hwmI, hwmIIb, hwmIIa
        public var shortLabel: String {
            switch self {
            case .unknown: return "Lage unklar"
            case .normal: return "Normal"
            case .hwmI: return "HWM I"
            case .hwmIIb: return "Sperre IIb"
            case .hwmIIa: return "Sperre IIa"
            }
        }
        public var detail: String {
            switch self {
            case .unknown: return "Basel-Pegel fehlt oder ist älter als 60 Minuten. Massgebend sind die Schweizerischen Rheinhäfen."
            case .normal: return "Unter Hochwassermarke I (< 700 cm)."
            case .hwmI: return "Voralarm ab 700 cm Pegel Basel-Rheinhalle."
            case .hwmIIb: return "Kleinschifffahrt und Fähren Basel–Rheinfelden gesperrt."
            case .hwmIIa: return "Schifffahrt Rheinfelden–Kembs gesperrt."
            }
        }
    }
    public static let maximumCurrentAge: TimeInterval = 3600
    public static let officialThresholdStages: [Stage] = [.hwmI, .hwmIIb, .hwmIIa]

    public static func stage(gaugeCentimetres: Double?) -> Stage {
        guard let gauge = gaugeCentimetres, gauge.isFinite else { return .unknown }
        // Conversion from 248.20 m yields 819.9999999999989 cm in binary floating point.
        // A 1e-9 cm arithmetic tolerance preserves the exact source threshold without visible rounding.
        let comparable = gauge + 1e-9
        if comparable >= 820 { return .hwmIIa }
        if comparable >= 790 { return .hwmIIb }
        if comparable >= 700 { return .hwmI }
        return .normal
    }
    public static func isCurrent(measurement: Date?, cacheUpdated: Date?, now: Date = Date()) -> Bool {
        guard let measurement, let cacheUpdated,
              measurement.timeIntervalSince1970 > 0, cacheUpdated.timeIntervalSince1970 > 0,
              now.timeIntervalSince1970 > 0 else { return false }
        let measurementAge = now.timeIntervalSince(measurement)
        let cacheAge = now.timeIntervalSince(cacheUpdated)
        return measurementAge.isFinite && cacheAge.isFinite
            && (0...maximumCurrentAge).contains(measurementAge)
            && (0...maximumCurrentAge).contains(cacheAge)
    }
    public static func currentStage(gaugeCentimetres: Double?, measurement: Date?, cacheUpdated: Date?, now: Date = Date()) -> Stage {
        isCurrent(measurement: measurement, cacheUpdated: cacheUpdated, now: now) ? stage(gaugeCentimetres: gaugeCentimetres) : .unknown
    }
    public static func thresholdGraphValue(stage: Stage, centimetres: Bool) -> Double? {
        let cm: Double
        switch stage {
        case .hwmI: cm = 700
        case .hwmIIb: cm = 790
        case .hwmIIa: cm = 820
        default: return nil
        }
        return centimetres ? cm : 240 + cm / 100
    }
}
